### PR TITLE
enable info collected during sut boot-time

### PR DESCRIPTION
**Pull request by ZhangMan1126**
manipluate executing action state
collect and restore objective info (such as targets and external service info) during the boot-time (only for JVM)
tests for checking whether the boot-time targets are covered with the exported targets file.

### CODE REVIEW

File: core/src/main/kotlin/org/evomaster/core/search/FitnessValue.kt [hunk 1]
```diff
@@ -146,6 +147,29 @@ class FitnessValue(
...
     }
 
+    /**
+     * this method is to report the union results with targets at boot-time
+     *
+     * @return first is boot-time (negative means that the boot-time info is unavailable)
+     *         second is search time, and
+     *         third is a union of boot-time and search time
+     */
+    fun unionWithBootTimeCoveredTargets(prefix: String?, idMapper: IdMapper, bootTimeInfoDto: BootTimeInfoDto?, unavailableBootTime: Int = -1): Triple<Int, Int, Int>{
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/FitnessValue.kt:157]: the use of `Triple` is fine in a local scope, but not when returned by a public method, as it makes code harder to read. eg., how to understand what `first` and `third` are referring to semantically.
create a new class with meaningful names to replace this `Triple`

Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/FitnessValue.kt:157]: the use of `Triple` is fine in a local scope, but not when returned by a public method, as it makes code harder to read. eg., how to understand what `first` and `third` are referring to semantically.
create a new class with meaningful names to replace this `Triple`

File: e2e-tests/spring-rest-openapi-v3/src/main/kotlin/com/foo/rest/examples/spring/openapi/v3/boottimetargets/BootTimeTargetsRest.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    private val startupTime  = LocalDateTime.now()
+    private var startupInfo : String
+
+    init {
+        startupInfo = "${branchTarget()}:$startupTime"
+        println("startupInfo:$startupInfo")
+    }
+
+    companion object{
+        // TODO might need a further check, targets for lines 24 and 25 seems to be skipped
```
Comment (arcuri82) [inline e2e-tests/spring-rest-openapi-v3/src/main/kotlin/com/foo/rest/examples/spring/openapi/v3/boottimetargets/BootTimeTargetsRest.kt:23]: yep, indeed, we do not instrument class initializers

Comment (arcuri82) [inline e2e-tests/spring-rest-openapi-v3/src/main/kotlin/com/foo/rest/examples/spring/openapi/v3/boottimetargets/BootTimeTargetsRest.kt:23]: yep, indeed, we do not instrument class initializers

**Review by arcuri82 — CHANGES_REQUESTED** [review]

**Review by seran — COMMENTED** [review]
Thank you @ZhangMan1126. I have no comments.

File: core/src/main/kotlin/org/evomaster/core/search/FitnessValue.kt [hunk 3]
```diff
@@ -167,7 +166,7 @@ class FitnessValue(
                     duplicatedcounter++
             }
         }
-        return Triple(bootTime.size, searchTime, bootTime.size + searchTime - duplicatedcounter)
+        return TargetStatisticTriple(bootTime.size, searchTime, bootTime.size + searchTime - duplicatedcounter)
```
[the remark this answers is not in the document]
Comment (ZhangMan1126) [reply core/src/main/kotlin/org/evomaster/core/search/FitnessValue.kt]: maybe it only happens for class due to class load? 
to ensure that we avoid the repeated ones, we might still need to have a check on them

File: core/src/main/kotlin/org/evomaster/core/search/TargetStatisticTriple.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+     */
+    val bootTime: Int,
+    /**
+     * achieved by search
+     */
+    val searchTime: Int,
+    /**
+     * a total of unique targets covered at the end
+     */
+    val total: Int
```
[the remark this answers is not in the document]
Comment (arcuri82) [reply core/src/main/kotlin/org/evomaster/core/search/TargetStatisticTriple.kt]: @ZhangMan1126 hi. thx. wondering if this is related to handling of `firstTimeEncountered`, which might not have been reset, maybe?

**Review by arcuri82 — APPROVED** [review]

**Pull request merged**